File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/RelationalDbDebeziumPropertiesManager.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.cdk.integrations.debezium.internals;

import com.fasterxml.jackson.databind.JsonNode;
import io.airbyte.cdk.db.jdbc.JdbcUtils;
import io.airbyte.protocol.models.v0.ConfiguredAirbyteCatalog;
import io.airbyte.protocol.models.v0.ConfiguredAirbyteStream;
import io.airbyte.protocol.models.v0.SyncMode;
import java.util.Iterator;
import java.util.Properties;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;
import org.codehaus.plexus.util.StringUtils;

public class RelationalDbDebeziumPropertiesManager extends DebeziumPropertiesManager {

  public RelationalDbDebeziumPropertiesManager(final Properties properties,
                                               final JsonNode config,
                                               final ConfiguredAirbyteCatalog catalog) {
    super(properties, config, catalog);
  }

  @Override
  protected Properties getConnectionConfiguration(JsonNode config) {
    final Properties properties = new Properties();

    // db connection configuration
    properties.setProperty("database.hostname", config.get(JdbcUtils.HOST_KEY).asText());
    properties.setProperty("database.port", config.get(JdbcUtils.PORT_KEY).asText());
    properties.setProperty("database.user", config.get(JdbcUtils.USERNAME_KEY).asText());
    properties.setProperty("database.dbname", config.get(JdbcUtils.DATABASE_KEY).asText());

    if (config.has(JdbcUtils.PASSWORD_KEY)) {
      properties.setProperty("database.password", config.get(JdbcUtils.PASSWORD_KEY).asText());
    }

    return properties;
  }

  @Override
  protected String getName(JsonNode config) {
    return config.get(JdbcUtils.DATABASE_KEY).asText();
  }

  @Override
  protected Properties getIncludeConfiguration(ConfiguredAirbyteCatalog catalog, JsonNode config) {
    final Properties properties = new Properties();

    // table selection
    properties.setProperty("table.include.list", getTableIncludelist(catalog));
    // column selection
    properties.setProperty("column.include.list", getColumnIncludeList(catalog));

    return properties;
  }

  public static String getTableIncludelist(final ConfiguredAirbyteCatalog catalog) {
    // Turn "stream": {
    // "namespace": "schema1"
    // "name": "table1
    // },
    // "stream": {
    // "namespace": "schema2"
    // "name": "table2
    // } -------> info "schema1.table1, schema2.table2"

    return catalog.getStreams().stream()
        .filter(s -> s.getSyncMode() == SyncMode.INCREMENTAL)
        .map(ConfiguredAirbyteStream::getStream)
        .map(stream -> stream.getNamespace() + "." + stream.getName())
        // debezium needs commas escaped to split properly
        .map(x -> StringUtils.escape(Pattern.quote(x), ",".toCharArray(), "\\,"))
        .collect(Collectors.joining(","));
  }

  public static String getColumnIncludeList(final ConfiguredAirbyteCatalog catalog) {
    // Turn "stream": {
    // "namespace": "schema1"
    // "name": "table1"
    // "jsonSchema": {
    // "properties": {
    // "column1": {
    // },
    // "column2": {
    // }
    // }
    // }
    // } -------> info "schema1.table1.(column1 | column2)"

    return catalog.getStreams().stream()
        .filter(s -> s.getSyncMode() == SyncMode.INCREMENTAL)
        .map(ConfiguredAirbyteStream::getStream)
        .map(s -> {
          final String fields = parseFields(s.getJsonSchema().get("properties").fieldNames());
          // schema.table.(col1|col2)
          return Pattern.quote(s.getNamespace() + "." + s.getName()) + (StringUtils.isNotBlank(fields) ? "\\." + fields : "");
        })
        .map(x -> StringUtils.escape(x, ",".toCharArray(), "\\,"))
        .collect(Collectors.joining(","));
  }

  private static String parseFields(final Iterator<String> fieldNames) {
    if (fieldNames == null || !fieldNames.hasNext()) {
      return "";
    }
    final Iterable<String> iter = () -> fieldNames;
    return StreamSupport.stream(iter.spliterator(), false)
        .map(f -> Pattern.quote(f))
        .collect(Collectors.joining("|", "(", ")"));
  }

}

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/integrations/debezium/internals/RelationalDbDebeziumPropertiesManager.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.integrations.debezium.internals

import com.fasterxml.jackson.databind.JsonNode
import io.airbyte.cdk.db.jdbc.JdbcUtils
import io.airbyte.protocol.models.v0.AirbyteStream
import io.airbyte.protocol.models.v0.ConfiguredAirbyteCatalog
import io.airbyte.protocol.models.v0.ConfiguredAirbyteStream
import io.airbyte.protocol.models.v0.SyncMode
import java.util.*
import java.util.regex.Pattern
import java.util.stream.Collectors
import java.util.stream.StreamSupport
import org.codehaus.plexus.util.StringUtils

class RelationalDbDebeziumPropertiesManager(
    properties: Properties,
    config: JsonNode,
    catalog: ConfiguredAirbyteCatalog,
    completedStreamNames: List<String>
) : DebeziumPropertiesManager(properties, config, catalog, completedStreamNames) {
    override fun getConnectionConfiguration(config: JsonNode): Properties {
        val properties = Properties()

        // db connection configuration
        properties.setProperty("database.hostname", config[JdbcUtils.HOST_KEY].asText())
        properties.setProperty("database.port", config[JdbcUtils.PORT_KEY].asText())
        properties.setProperty("database.user", config[JdbcUtils.USERNAME_KEY].asText())
        properties.setProperty("database.dbname", config[JdbcUtils.DATABASE_KEY].asText())

        if (config.has(JdbcUtils.PASSWORD_KEY)) {
            properties.setProperty("database.password", config[JdbcUtils.PASSWORD_KEY].asText())
        }

        return properties
    }

    override fun getName(config: JsonNode): String {
        return config[JdbcUtils.DATABASE_KEY].asText()
    }

    override fun getIncludeConfiguration(
        catalog: ConfiguredAirbyteCatalog,
        config: JsonNode?,
        streamNames: List<String>
    ): Properties {
        val properties = Properties()

        // table selection
        properties.setProperty("table.include.list", getTableIncludelist(catalog, streamNames))
        // column selection
        properties.setProperty("column.include.list", getColumnIncludeList(catalog, streamNames))

        return properties
    }

    companion object {
        fun getTableIncludelist(
            catalog: ConfiguredAirbyteCatalog,
            completedStreamNames: List<String>
        ): String {
            // Turn "stream": {
            // "namespace": "schema1"
            // "name": "table1
            // },
            // "stream": {
            // "namespace": "schema2"
            // "name": "table2
            // } -------> info "schema1.table1, schema2.table2"

            return catalog.streams
                .filter { s: ConfiguredAirbyteStream -> s.syncMode == SyncMode.INCREMENTAL }
                .map { obj: ConfiguredAirbyteStream -> obj.stream }
                .map { stream: AirbyteStream -> stream.namespace + "." + stream.name }
                .filter { streamName: String -> completedStreamNames.contains(streamName) }
                // debezium needs commas escaped to split properly
                .joinToString(",") { x: String ->
                    StringUtils.escape(Pattern.quote(x), ",".toCharArray(), "\\,")
                }
        }

        fun getColumnIncludeList(
            catalog: ConfiguredAirbyteCatalog,
            completedStreamNames: List<String>
        ): String {
            // Turn "stream": {
            // "namespace": "schema1"
            // "name": "table1"
            // "jsonSchema": {
            // "properties": {
            // "column1": {
            // },
            // "column2": {
            // }
            // }
            // }
            // } -------> info "schema1.table1.(column1 | column2)"

            return catalog.streams
                .filter { s: ConfiguredAirbyteStream -> s.syncMode == SyncMode.INCREMENTAL }
                .map { obj: ConfiguredAirbyteStream -> obj.stream }
                .filter { stream: AirbyteStream ->
                    completedStreamNames.contains(stream.namespace + "." + stream.name)
                }
                .map { s: AirbyteStream ->
                    val fields = parseFields(s.jsonSchema["properties"].fieldNames())
                    Pattern.quote(s.namespace + "." + s.name) +
                        (if (StringUtils.isNotBlank(fields)) "\\.$fields" else "")
                }
                .joinToString(",") { x: String -> StringUtils.escape(x, ",".toCharArray(), "\\,") }
        }

        private fun parseFields(fieldNames: Iterator<String>?): String {
            if (fieldNames == null || !fieldNames.hasNext()) {
                return ""
            }
            val iter = Iterable { fieldNames }
            return StreamSupport.stream(iter.spliterator(), false)
                .map { f: String -> Pattern.quote(f) }
                .collect(Collectors.joining("|", "(", ")"))
        }
    }
}
